src/main/scala/mesosphere/marathon/core/storage/IdResolver.scala

```diff
@@ -0,0 +1,40 @@
+package mesosphere.marathon.core.storage
+
+import java.time.OffsetDateTime
+
+/**
+  * Resolver for Marathon Internal Persistence IDs (by `Key Type`, `Value Type` and `Serialized Type`)
+  * For example, Applications may be identified by the string "/apps/abc" and stored internally at
+  * "/apps/a/a58ec17d-735d-4c3f-9aa8-d44c764aa31b"
+  *
+  * This IdResolver must be provided for all storage engines. See [[mesosphere.marathon.core.storage.PersistenceStore]]
+  *
+  * @tparam Id         The marathon type of the ID for the given Value type
+  * @tparam K          The persisted type of the ID
+  * @tparam Category   The category that 'V' belongs to.
+  * @tparam V          The value type being stored
+  */
+trait IdResolver[Id, V, Category, K] {
+  /**
+    * Translate the marathon id into the given persisted format
+    */
+  def toStorageId(id: Id, version: Option[OffsetDateTime]): K
```

Comment:
OffsetDateTime is part of Java8 Time and it has a particular advantage in utilizing the timezone while still being in UTC. Java8 Time is essentially joda2.